data/pigui/modules/radar.lua

```diff
@@ -95,144 +172,222 @@ local function display2DRadar(cntr, size)
+				toggle_radar = true
 			end
 			if ui.selectable(lui.HUD_3D_RADAR, not shouldDisplay2DRadar, {}) then
-				Event.Queue('onChangeMFD', 'scanner')
+				toggle_radar = true
 			end
 		end)
-		if shouldDisplay2DRadar then
-			display2DRadar(cntr, size)
-		else
-			display3DRadar(cntr, Vector2(ui.reticuleCircleRadius * 1.8, size * 2))
+
+		if ui.isMouseClicked(1) then
+			click_on_radar = true
+		end
+		if not toggle_radar and click_on_radar and ui.isMouseReleased(1) then
+			ui.openPopup("radarselector")
+		end
+		-- TODO: figure out how to "capture" the mouse wheel to prevent
+		-- the game engine from using it to also zoom the viewport
+		if zoom == 0 then
+			zoom = ui.getMouseWheel()
 		end
 	end
-end
 
-Event.Register('onChangeMFD', function(selected)
-	shouldDisplay2DRadar = selected == "radar";
+	-- Update the zoom level
+	if zoom > 0 then
+		instrument:zoomIn()
+	elseif zoom < 0 then
+		instrument:zoomOut()
+	end
+
+	-- Draw the actual radar - this can't be in a window or the 3D scanner background doesn't render
+	instrument:draw(center)
+
+	-- Draw the radar buttons and info - this needs to be in a window, otherwise the buttons don't work.
+	local window_width = ui.reticuleCircleRadius * 1.8
+	local window_height = radar2d.size / 3.5
+	local window_pos = Vector2(center.x - window_width / 2, center.y + radar2d.size - window_height - SCREEN_BORDER)
+	local windowFlags = ui.WindowFlags {"NoTitleBar", "NoResize", "NoFocusOnAppearing", "NoBringToFrontOnFocus", "NoSavedSettings"}
+	ui.setNextWindowPos(window_pos, "Always")
+	ui.setNextWindowPadding(Vector2(0))
+	ui.setNextWindowSize(Vector2(window_width, window_height), "Always")
+
+	ui.window("radar_buttons", windowFlags, function()
+
+		-- Draw radar mode toggle button
+		local icon = shouldDisplay2DRadar and radar3d.icon or radar2d.icon
+		local clicked = ui.mainMenuButton(icon, lui.HUD_RADAR_TOGGLE_MODE, false, Vector2(window_height))
+		if toggle_radar or clicked then
+			shouldDisplay2DRadar = not shouldDisplay2DRadar
+		end
+
+		-- Draw zoom mode indicator
+		if not shouldDisplay2DRadar then
+			local button_size = window_height / 1.5
+			local tt = radar3d.auto_zoom and lui.HUD_RADAR_ZOOM_MODE_AUTOMATIC or lui.HUD_RADAR_ZOOM_MODE_MANUAL
+			ui.sameLine()
+			ui.addCursorPos(Vector2(0, window_height - button_size))
+			icon = instrument:isAutoZoom() and icons.radar_automatic or icons.radar_manual
+			ui.mainMenuButton(icon, tt, ui.theme.buttonColors.disabled, Vector2(button_size))
+		end
+
+		-- Draw radar range
+		local distance = ui.Format.Distance(instrument:getZoom())
+		local textpos = ui.getWindowPos() + Vector2(window_width, window_height)
+		ui.addStyledText(textpos, ui.anchor.right, ui.anchor.bottom, distance, colors.frame, pionillium.small, lui.HUD_RADAR_DISTANCE, colors.lightBlackBackground)
+
+	end) -- window
+
+end -- function displayRadar()
+
+-- enable radar input frame on game start
+Event.Register("onGameStart", function()
+	input_frame:AddToStack()
```

Comment:
Let me take a look at how to do this..

Reply:
Fixed in 1b370e33f984658232158bddafb4e4840b8a9df1

Reply:
Note: `InputFrame:RemoveFromStack()` "should" be idempotent - you can instead do `if currentView ~= "world" then frame:RemoveFromStack() end`.

If it's not idempotent, bug me about it and I'll fix that.

Reply:
That's less performant though calling into `RemoveFromStack()` on every view switch instead of only when necessary..  `RemoveFromStack` is not only a call into the C++, but it then also to look through the entire vector.

Reply:
While I do agree, I consider this to be only a stopgap solution, and a few microseconds to walk a vector of pointers (with N < 30 on average) once every 5-300 seconds is not a particular performance issue in my mind. I'm much more concerned about introducing new API which will likely wind up completely unneeded, especially if we add support for opening a View "inside" another View as a floating window.